p4_symbolic/ir/ir.proto

```diff
@@ -0,0 +1,230 @@
+  // Essentially, this is the action signature.
+  pdpi.ir.IrActionDefinition action_definition = 1;
+  // Action implementation, essentially its body.
+  ActionImplementation action_implementation = 2;
+}
+
+// Action implementation details.
+message ActionImplementation {
+  // List of statements in action body in sequence.
+  repeated Statement action_body = 1;
+}
+
+// Overall table structure, combining definition with implementation.
+message Table {
+  // This contains the name, id, action list, field matches, and max size.
+  pdpi.ir.IrTableDefinition table_definition = 1;
+  // This contains table implementation details.
+  TableImplementation table_implementation = 2;
+}
+
+// Table implementation details that are not exposed in the definition.
+message TableImplementation {
+  // The overal type of the match(s) of the table.
+  // In case the table have several keys with different match type,
+  // the overall match type of the table would be the most general.
+  // There is always a most general match because of P4 restrications
+  // on match combination (which may be under-documented).
+  // See https://github.com/p4lang/p4-spec/issues/411
+  p4_symbolic.bmv2.TableMatchTypeEnum match_type = 5;
+  // Table type with respect to action selection.
+  p4_symbolic.bmv2.TableTypeEnum type = 6;
+}
+
+// An abstract p4 statement corresponding to a top level operation within
+// an action body.
+message Statement {
+  p4_symbolic.bmv2.SourceLocation source_info = 1;
+  // Various statement concrete types.
+  oneof statement {
+    AssignmentStatement assignment = 2;
+  }
+}
+
+// An assignment statement of the form <left> = <right>.
+// Note that left can be an rvalue expression.
+message AssignmentStatement {
+  LValue left = 1;
+  RValue right = 2;
+}
+
+// The structure of an abstract RValue.
+message RValue {
+  oneof rvalue {
+    FieldValue field_value = 2;
+    HexstrValue hexstr_value = 3;
+    BoolValue bool_value = 4;
+    StringValue string_value = 5;
+    Variable variable_value = 6;
+    RExpression expression_value = 7;
+  }
+}
+
+// The structure of an abstract LValue.
+message LValue {
+  oneof lvalue {
+    FieldValue field_value = 1;
+    Variable variable_value = 2;
+  }
+}
+
+// A header field accessed statically.
+// Field access that is dynamic (either in the index/offset or in the header
+// object) is defined under AccessExpression below.
+message FieldValue {
+  string header_name = 1;
+  string field_name = 2;
+}
+
+// Corresponds to a hex bigendiean encoding of a bit-string.
+// In the bmv2 JSON format. The hexstring can be prefixed by "-"
+// for negative values.
+// We remove any such prefix and explicitly set the negative flag.
+message HexstrValue {
+  string value = 1;
+  bool negative = 2;
+}
+
+// A boolean literal value.
+message BoolValue {
+  bool value = 1;
+}
+
+// A string literal value.
+message StringValue {
+  string value = 1;
+}
+
+// Our explicit representation of a runtime variable.
```

Comment:
No need to compare to the other format, just say something along the lines of "Variable name". If we every change the phase 1 to some other front-end, this will all be stale.